Explore/Pyroscope: Empty result in Pyroscope explore crashes grafana

### What happened?

Grafana server crashes if a profile explore query goes too far into the past:

 ```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x9847271]

goroutine 16853 [running]:
github.com/grafana/grafana/pkg/tsdb/grafana-pyroscope-datasource.(*PyroscopeClient).GetProfile(0xc00540e8d0, {0x21cf358, 0xc004366460}, {0xc004ef85d0, 0x2b}, {0xc00634f6e0, 0x18}, 0x189c5f729d0, 0x189cb1d85d0, 0x0)
        github.com/grafana/grafana/pkg/tsdb/grafana-pyroscope-datasource/pyroscopeClient.go:152 +0x171
github.com/grafana/grafana/pkg/tsdb/grafana-pyroscope-datasource.(*PyroscopeDatasource).query.func2()
        github.com/grafana/grafana/pkg/tsdb/grafana-pyroscope-datasource/query.go:89 +0x21e
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.3.0/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync@v0.3.0/errgroup/errgroup.go:72 +0xa5
```

### What did you expect to happen?

When there is no search result Grafana should display that instead of crash

### Did this work before?

It may have.

### How do we reproduce it?

1. Select a Pyroscope data source in explore
2. Query beyond the retention window (30days)
3. Grafana restart after the panic above


### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana v10.2.0-61567 (e7c1e5e4c9) (admin-dev* as of creation of this issue)



### Grafana platform?

Kubernetes

### Datasource(s)?

Pyroscope

--- a/pkg/tsdb/grafana-pyroscope-datasource/pyroscopeClient.go
+++ b/pkg/tsdb/grafana-pyroscope-datasource/pyroscopeClient.go
@@ -149,6 +149,11 @@ func (c *PyroscopeClient) GetProfile(ctx context.Context, profileTypeID, labelSe
 		return nil, err
 	}
 
+	if resp.Msg.Flamegraph == nil {
+		// Not an error, can happen when querying data oout of range.
+		return nil, nil
+	}
+
 	levels := make([]*Level, len(resp.Msg.Flamegraph.Levels))
 	for i, level := range resp.Msg.Flamegraph.Levels {
 		levels[i] = &Level{

--- a/pkg/tsdb/grafana-pyroscope-datasource/pyroscopeClient_test.go
+++ b/pkg/tsdb/grafana-pyroscope-datasource/pyroscopeClient_test.go
@@ -51,10 +51,21 @@ func Test_PyroscopeClient(t *testing.T) {
 		}
 		require.Equal(t, series, resp)
 	})
+
+	t.Run("GetProfile with empty response", func(t *testing.T) {
+		connectClient.SendEmptyProfileResponse = true
+		maxNodes := int64(-1)
+		resp, err := client.GetProfile(context.Background(), "memory:alloc_objects:count:space:bytes", "{}", 0, 100, &maxNodes)
+		require.Nil(t, err)
+		// Mainly ensuring this does not panic like before
+		require.Nil(t, resp)
+		connectClient.SendEmptyProfileResponse = false
+	})
 }
 
 type FakePyroscopeConnectClient struct {
-	Req any
+	Req                      any
+	SendEmptyProfileResponse bool
 }
 
 func (f *FakePyroscopeConnectClient) ProfileTypes(ctx context.Context, c *connect.Request[querierv1.ProfileTypesRequest]) (*connect.Response[querierv1.ProfileTypesResponse], error) {
@@ -75,6 +86,9 @@ func (f *FakePyroscopeConnectClient) Series(ctx context.Context, c *connect.Requ
 
 func (f *FakePyroscopeConnectClient) SelectMergeStacktraces(ctx context.Context, c *connect.Request[querierv1.SelectMergeStacktracesRequest]) (*connect.Response[querierv1.SelectMergeStacktracesResponse], error) {
 	f.Req = c
+	if f.SendEmptyProfileResponse {
+		return &connect.Response[querierv1.SelectMergeStacktracesResponse]{Msg: &querierv1.SelectMergeStacktracesResponse{}}, nil
+	}
 	return &connect.Response[querierv1.SelectMergeStacktracesResponse]{
 		Msg: &querierv1.SelectMergeStacktracesResponse{
 			Flamegraph: &querierv1.FlameGraph{

--- a/pkg/tsdb/grafana-pyroscope-datasource/query.go
+++ b/pkg/tsdb/grafana-pyroscope-datasource/query.go
@@ -91,22 +91,30 @@ func (d *PyroscopeDatasource) query(ctx context.Context, pCtx backend.PluginCont
 				logger.Error("Error GetProfile()", "err", err)
 				return err
 			}
-			frame := responseToDataFrames(prof)
+
+			var frame *data.Frame
+			if prof != nil {
+				frame := responseToDataFrames(prof)
+
+				// If query called with streaming on then return a channel
+				// to subscribe on a client-side and consume updates from a plugin.
+				// Feel free to remove this if you don't need streaming for your datasource.
+				if qm.WithStreaming {
+					channel := live.Channel{
+						Scope:     live.ScopeDatasource,
+						Namespace: pCtx.DataSourceInstanceSettings.UID,
+						Path:      "stream",
+					}
+					frame.SetMeta(&data.FrameMeta{Channel: channel.String()})
+				}
+			} else {
+				// We still send empty data frame to give feedback that query really run, just didn't return any data.
+				frame = getEmptyDataFrame()
+			}
 			responseMutex.Lock()
 			response.Frames = append(response.Frames, frame)
 			responseMutex.Unlock()
 
-			// If query called with streaming on then return a channel
-			// to subscribe on a client-side and consume updates from a plugin.
-			// Feel free to remove this if you don't need streaming for your datasource.
-			if qm.WithStreaming {
-				channel := live.Channel{
-					Scope:     live.ScopeDatasource,
-					Namespace: pCtx.DataSourceInstanceSettings.UID,
-					Path:      "stream",
-				}
-				frame.SetMeta(&data.FrameMeta{Channel: channel.String()})
-			}
 			return nil
 		})
 	}
@@ -271,6 +279,18 @@ func (pt *ProfileTree) String() string {
 		}
 	}
 	return tree.String()
+}
+
+func getEmptyDataFrame() *data.Frame {
+	var emptyProfileDataFrame = data.NewFrame("response")
+	emptyProfileDataFrame.Meta = &data.FrameMeta{PreferredVisualization: "flamegraph"}
+	emptyProfileDataFrame.Fields = data.Fields{
+		data.NewField("level", nil, []int64{}),
+		data.NewField("value", nil, []int64{}),
+		data.NewField("self", nil, []int64{}),
+		data.NewField("label", nil, []string{}),
+	}
+	return emptyProfileDataFrame
 }
 
 type CustomMeta struct {

--- a/pkg/tsdb/grafana-pyroscope-datasource/query.go
+++ b/pkg/tsdb/grafana-pyroscope-datasource/query.go
@@ -94,7 +94,7 @@ func (d *PyroscopeDatasource) query(ctx context.Context, pCtx backend.PluginCont
 
 			var frame *data.Frame
 			if prof != nil {
-				frame := responseToDataFrames(prof)
+				frame = responseToDataFrames(prof)
 
 				// If query called with streaming on then return a channel
 				// to subscribe on a client-side and consume updates from a plugin.

